File: go/sign.go
package vitni

import (
	"crypto/ed25519"
	"encoding/base64"
	"encoding/json"
	"errors"

	"github.com/gowebpki/jcs"
)

// Version is the Vitni protocol version string stamped into receipts.
const Version = "vitni/0.2"

// MCPMetaKey is the MCP `_meta` key under which a Vitni receipt is carried.
const MCPMetaKey = "dev.vitni/receipt"

// Cost is the cost block of a receipt. Magnitudes are decimal STRINGS (the 2^53
// defense — never JSON numbers). RailRef is an optional reference string.
type Cost struct {
	Tokens    string  `json:"tokens"`
	USDMicros string  `json:"usd_micros"`
	WallMs    string  `json:"wall_ms"`
	RailRef   *string `json:"rail_ref"`
}

// Receipt is a Vitni receipt payload (without receipt_id; that is the content
// address derived from JCS(receipt)). All nullable fields use pointers so they
// serialize as JSON null when unset.
type Receipt struct {
	V                 string  `json:"v"`
	Binding           string  `json:"binding"`
	ActionRef         *string `json:"action_ref"`
	PerformerID       string  `json:"performer_id"`
	RequesterID       *string `json:"requester_id"`
	Method            string  `json:"method"`
	InputsHash        string  `json:"inputs_hash"`
	OutputsHash       string  `json:"outputs_hash"`
	Cost              Cost    `json:"cost"`
	Status            string  `json:"status"`
	Reason            *string `json:"reason"`
	ParentReceiptHash *string `json:"parent_receipt_hash"`
	ParentPerformerID *string `json:"parent_performer_id,omitempty"`
	LogPolicy         string  `json:"log_policy"`
	Ts                string  `json:"ts"`
	Nonce             string  `json:"nonce"`
}

// Sign produces a JWS-compact signed receipt (header.payload.signature) over the
// JCS-canonical octets of the receipt, using Ed25519 (alg "EdDSA").
//
// The payload is the RFC 8785 canonical form of the receipt, so a verifier's
// non_canonical_payload check passes. The receipt's version field is forced to
// the current Version constant. The signing input is the ASCII bytes
// b64url(header).b64url(payload), matching Verify.
func Sign(r Receipt, kid string, priv ed25519.PrivateKey) (string, error) {
	if len(priv) != ed25519.PrivateKeySize {
		return "", errors.New("invalid ed25519 private key size")
	}
	if kid == "" {
		return "", errors.New("kid must not be empty")
	}

	// Stamp the protocol version.
	r.V = Version

	// Canonicalize the payload (JCS) so it round-trips with Verify step 7.
	payloadJSON, err := json.Marshal(r)
	if err != nil {
		return "", err
	}
	payloadCanon, err := jcs.Transform(payloadJSON)
	if err != nil {
		return "", err
	}

	// Canonical JWS protected header: {"alg":"EdDSA","kid":<kid>}.
	headerJSON, err := json.Marshal(map[string]string{"alg": "EdDSA", "kid": kid})
	if err != nil {
		return "", err
	}
	headerCanon, err := jcs.Transform(headerJSON)
	if err != nil {
		return "", err
	}

	headerSeg := base64.RawURLEncoding.EncodeToString(headerCanon)
	payloadSeg := base64.RawURLEncoding.EncodeToString(payloadCanon)
	signingInput := []byte(headerSeg + "." + payloadSeg)

	sig := ed25519.Sign(priv, signingInput)
	sigSeg := base64.RawURLEncoding.EncodeToString(sig)

	return headerSeg + "." + payloadSeg + "." + sigSeg, nil
}

File: ts/src/commands/sign.ts
/**
 * sign command — produce a deterministic EdDSA JWS signed receipt (conformance §10).
 *
 * Input:
 *   { "receipt": <Receipt WITHOUT receipt_id>, "kid": "<string>",
 *     "private_key_b64": "<base64 of a 32-byte Ed25519 RFC 8032 seed>" }
 * Output:
 *   { "signed_receipt": "<b64url(header).b64url(payload).b64url(sig)>" }
 *
 * The signing input and header/payload framing match the §7 verifier and the
 * MCP middleware signer exactly, so a produced receipt round-trips through verify
 * with no non_canonical_payload failure.
 *
 * KEY ENCODING (load-bearing): private_key_b64 is the 32-byte RFC 8032 seed — NOT
 * the 64-byte expanded key. The seed is the only runtime-portable form. node:crypto
 * has no direct "seed" import, so we wrap the seed into a PKCS8 DER by prefixing the
 * fixed 16-byte Ed25519 PKCS8 header, then importing it.
 */
import { createPrivateKey, sign as cryptoSign, KeyObject } from 'node:crypto';
import { decodeStdBase64 } from './decode.js';
import { jcsBytes, jcsString } from './jcs.js';
import { VERSION } from '../version.js';

export interface SignInput {
  receipt: Record<string, unknown>;
  kid: string;
  private_key_b64: string;
}

export interface SignOutput {
  signed_receipt: string;
}

/** The fixed 16-byte ASN.1/PKCS8 prefix for a raw Ed25519 private seed (RFC 8410). */
const PKCS8_ED25519_PREFIX = Buffer.from('302e020100300506032b657004220420', 'hex');

/** Closed set of top-level Receipt fields (§2); anything else is rejected at sign. */
const KNOWN_RECEIPT_KEYS = new Set([
  'v', 'binding', 'action_ref', 'performer_id', 'requester_id', 'method',
  'inputs_hash', 'outputs_hash', 'cost', 'status', 'reason',
  'parent_receipt_hash', 'parent_performer_id', 'log_policy', 'ts', 'nonce', 'ext',
]);

function b64url(buf: Buffer): string {
  return buf.toString('base64url');
}

/**
 * Decode base64, accepting BOTH standard and raw (unpadded) base64 — the same
 * strict-canonical accept set as the `digest` command (and Go's
 * vitni.DecodeStdBase64). One decoder repo-wide keeps sign/keygen seeds and
 * receipt byte-sources byte-identical across impls.
 */
export function decodeFlexibleBase64(s: string): Buffer | null {
  if (typeof s !== 'string') return null;
  return decodeStdBase64(s);
}

/** Wrap a 32-byte Ed25519 seed into a node KeyObject. Returns null on failure. */
export function privateKeyFromSeed(seed: Buffer): KeyObject | null {
  if (seed.length !== 32) return null;
  try {
    const der = Buffer.concat([PKCS8_ED25519_PREFIX, seed]);
    return createPrivateKey({ key: der, format: 'der', type: 'pkcs8' });
  } catch {
    return null;
  }
}

export function sign(input: SignInput): SignOutput | { error: string } {
  // --- validate input shape ---
  if (input === null || typeof input !== 'object') {
    return { error: 'invalid_input' };
  }
  const receipt = input.receipt;
  if (receipt === null || typeof receipt !== 'object' || Array.isArray(receipt)) {
    return { error: 'invalid_input' };
  }

  // --- receipt_id must be absent (reuse the existing contract) ---
  if (Object.prototype.hasOwnProperty.call(receipt, 'receipt_id')) {
    return { error: 'receipt_id_must_be_absent' };
  }

  // --- reject unknown top-level keys (§2): the Receipt is a defined object;
  // forward-compatible extensions belong in `ext`. Signing over an unknown key
  // would let un-schema'd data into the signed bytes and receipt_id. ---
  for (const k of Object.keys(receipt)) {
    if (!KNOWN_RECEIPT_KEYS.has(k)) {
      return { error: 'invalid_input' };
    }
  }

  // --- cost magnitudes MUST be decimal strings (§4.1): reject a numeric (or
  // any non-string) magnitude, matching Go's typed-struct rejection. ---
  const cost = receipt.cost;
  if (cost !== null && typeof cost === 'object' && !Array.isArray(cost)) {
    for (const field of ['tokens', 'usd_micros', 'wall_ms'] as const) {
      const val = (cost as Record<string, unknown>)[field];
      if (val !== undefined && typeof val !== 'string') {
        return { error: 'invalid_input' };
      }
    }
  }

  // --- kid required ---
  const kid = input.kid;
  if (typeof kid !== 'string' || kid === '') {
    return { error: 'kid_required' };
  }

  // --- decode + validate the 32-byte seed ---
  if (typeof input.private_key_b64 !== 'string') {
    return { error: 'invalid_private_key' };
  }
  const seed = decodeFlexibleBase64(input.private_key_b64);
  if (seed === null || seed.length !== 32) {
    return { error: 'invalid_private_key' };
  }
  const privateKey = privateKeyFromSeed(seed);
  if (privateKey === null) {
    return { error: 'invalid_private_key' };
  }

  // --- stamp protocol version (mirrors vitni.Sign / the middleware) ---
  const payload = { ...receipt, v: VERSION };

  // --- frame: canonical header + JCS-canonical payload, EdDSA over the signing input ---
  const header = { alg: 'EdDSA', kid };
  const headerSeg = b64url(Buffer.from(jcsString(header), 'utf8'));
  const payloadSeg = b64url(jcsBytes(payload));
  const signingInput = Buffer.from(`${headerSeg}.${payloadSeg}`, 'ascii');
  const sig = cryptoSign(null, signingInput, privateKey);
  const signed_receipt = `${headerSeg}.${payloadSeg}.${b64url(sig)}`;

  return { signed_receipt };
}
